IOSDOX-12139 Update SDK to Swift and made into Swift Package

ronnie3786 opened the pull request:
Jira Story: https://doximity.atlassian.net/browse/IOSDOX-12139

## Overview
I converted the legacy SDK from Objective-C to Swift and converted this to be a Swift Package to be used with SPM. I also kept the ability for users to do a manual integration and added the necessary Objc helper functions to keep this compatible with Objective-C.

## Technical Notes
Changes made:
- Converted the code to Swift
- Added Objective-C bridging
- Setup Swift Package details
- Removed Cocoapods and Carthage support (and mention of the README
- Added Unit Tests (updated implementation to accept a `MockApplication`
- Updated `gitignore` to ignore some build and local files

## Dev (as QA) Instructions
- Create a new Xcode Project for a basic iOS App
- Follow the SPM integration steps and usage from the updated README (for testing, you should be able to specify this branch name)
- Test the `DoxDialerCaller.shared.dialerIcon` and `DoxDialerCaller.shared.dialPhoneNumber("4254443333")` functionality
- [ ] Ensure the icon is visible as expected and the call will correctly deep-link into our Doximity app (if installed) or link to the App Store page
- Remove the SPM integration
- Follow the Manual Integration steps
- [ ] Ensure the manual integration works as expected and the same as the SPM integration (minus the need for the `import`)
- Bonus: Test the Objective-C integration

### Integration Demo

Here's a demo showing the Swift integration using SPM and by manually dragging in the two files noted in the `README`.

https://github.com/doximity/CallWithDoxDialer/assets/3907680/3559c7ff-77a3-417b-a63d-ddae368b2cfa

ronnie3786 committed:
Converted project to SPM package

ronnie3786 committed:
Setting resources and platforms properties to the package

ronnie3786 committed:
Added Objective-C support, fixed url path

Updating gitignore to ignore build generated data

ronnie3786 committed:
Added manual integration support and documentation

ronnie3786 committed:
Updated the README

ronnie3786 committed:
Added unit tests

ronnie3786 commented:
@dmallory I added you as a reviewer. Can you provide an extra set of eyes on the SPM setup and integration? Let me know if there are any improvements that can be made.

ronnie3786 commented on Sources/CallWithDoxDialer/CallWithDoxDialer.swift:
```diff
@@ -0,0 +1,117 @@
+import Foundation
+#if canImport(UIKit)
```
This wasn't required for integration into another project, but I couldn't get `swift build` to pass validation unless I added this check.

dmallory replied:
This feels to me like it's trying to remain valid on other platforms somewhere (MacOS)? but not worth a ton of time to figure out if it's working fine. I saw that it has `iOS` as the platform in the package config. 🤷🏼‍♂️

ronnie3786 commented on Sources/CallWithDoxDialer/CallWithDoxDialer.swift:
```diff
@@ -0,0 +1,117 @@
+
+private extension DoxDialerCaller {
+    /// Doximity icon asset from the swift package resources
+    var iconFromPackage: UIImage? {
+    #if SWIFT_PACKAGE
+        guard
+            let image = UIImage(
+                named: "icon",
+                in: .module,
+                compatibleWith: nil
+            )
+        else {
+            return UIImage()
+        }
+        return image
+    #else
+        return nil
+    #endif
+    }
```
This `#if SWIFT_PACKAGE` was needed when I added the manual integration option. The icon is stored in the Package Bundle (`.module`) but that bundle doesn't exist outside of the Package and would create a compile error when pulling in the files manually.

Manual integration uses the `var iconFromBundle: UIImage?` alternate.

dmallory commented on .gitignore:
```diff
@@ -17,7 +17,14 @@ profile
 Carthage/Checkouts/
-Carthage/Build/
\ No newline at end of file
+Carthage/Build/
```
Probably could just delete this one, makes it look like we're using `Carthage`.

dmallory commented on Package.swift:
```diff
@@ -0,0 +1,30 @@
+import PackageDescription
+
+let package = Package(
+    name: "CallWithDoxDialer",
```
For Comms to decide, but if you wanted to pick a different/better name for this project and also use it for the package, now would be a good time. The newer Ads SDK is `ios-ads-sdk` and the package is `DoximityAdsSDK` so something like `DoximityCallWithDialer` (if people really like the current name) or `DoximityDialerSDK` (which give more room for future additions) might follow that.

Chandlerdea replied:
I agree that we should update the name to be more inline with the naming convention of our current SDKs.

ronnie3786 replied:
@dmallory I updated the SDK to be renamed to `DoximityDialerSDK` in this commit: https://github.com/doximity/CallWithDoxDialer/pull/22/commits/a93363a3978cd16c28abdf8c696b5c496c588c38

dmallory commented on Sources/CallWithDoxDialer/CallWithDoxDialer.swift:
```diff
@@ -0,0 +1,117 @@
+import UIKit
+
+/// A struct to handle dialing phone numbers using the Doximity app.
+public struct DoxDialerCaller {
```
Similar to the package name, if you want to follow the newer Ads SDK they made a product decision that we want to spell out `Doximity` as much as possible and get away from `Dox` in any public APIs. The types there are spelled out like https://github.com/doximity/ios-ads-sdk/blob/develop/DoximityAdsSDK/Sources/DoximityAdsPortal.swift if you want to follow it.

dmallory reviewed (approved):
Functionality looks good to me, but worth thinking about naming as that's harder to keep changing in a public API.

Also should make sure this bridging does really work in an Objective-C project. If you want to easily keep that maintained you could include a simple Objective-C example app target as well, but at least that we've tested it initially.

Chandlerdea commented on Sources/CallWithDoxDialer/CallWithDoxDialer.swift:
```diff
@@ -0,0 +1,117 @@
+
+    /// Returns the Doximity icon image.
+    public var doximityIcon: UIImage {
+        return iconFromPackage ?? iconFromBundle ?? UIImage()
```
Should we throw an error her and let the consumers handle the case where we can't load the icon from the bundle?

Chandlerdea reviewed (approved):
Looks great, just one question but not a blocker.

ronnie3786 committed:
PR Feedback - code clean up

- Removing carthage from gitignore
- Remove #if canImport(UIKit) checks
- Throwing error for Dox UIImage functions

ronnie3786 committed:
Rename swift package to DoximityDialerSDK

Rename unit test file

Updating API to use DoximityDialer and remove “SDK”

ronnie3786 committed:
Moving OpenApplicationURL protocol to SDK file for manual integration

Chandlerdea reviewed (approved):
👍
